devise を使ってユーザー認証を実装する

masamune17 opened the pull request.

masamune17 committed:
deviceのセットアップ完了

masamune17 committed:
device の view 作成[C

masamune17 committed:
ユーザー登録フォームに郵便番号、住所、自己紹介を追加

masamune17 committed:
gem 'devise-i18n' をインストール

masamune17 committed:
deviceの日本語化対応、アプリ使用事にログイン必須化

ja.ymlにviews/device の各ファイルの日本語設定記載
translation_ja.yml にモデル名と属性を追加
application_controller.rb にbefore_action :authenticate_user! 追加

masamune17 committed:
リファクタリング

masamune17 committed:
Revert "リファクタリング"

This reverts commit 58ac0f959246c3f909fd7032726b383b56cab748.

masamune17 committed:
パスワード再設定時のメール送信機能追加

送信したメールは http://localhost:3000/letter_opener で確認

masamune17 committed:
ユーザーの一覧画面と詳細画面を作成

masamune17 committed:
ユーザー詳細画面の編集と日本語化

masamune17 committed:
ユーザーの一覧画面のページング処理、日本語化、リファクタリング

'#frozen_string_literal: true'のマジックコメントを必要箇所に追加
db/seeds.rb に テスト用の初期ユーザーデータを追加

masamune17 committed:
リファクタリング

masamune17 committed:
ユーザー情報の変更で郵便峰号、住所、自己紹介の変更が反映されるように修正

masamune17 committed:
郵便番号、住所、自己紹介のバリデーションを無効化

masamune17 committed:
郵番号、住所、自己紹介をサインアップ、更新時に反映されるように変更

masamune17 committed:
サインイン後のリダイレクト先をトップ画面に設定

masamune17 committed:
サインアップ後のリダイレクト先をトップ画面に設定

郵便番号、住所、自己紹介のバリテーション無効化

betachelsea commented on app/controllers/application_controller.rb:
```diff
@@ -1,4 +1,22 @@
+  protected
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:postcode])
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:address_city])
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:self_introduction])
```
一括で次のようにまとめて書けます！ 

```ruby
devise_parameter_sanitizer.permit(:sign_up, keys: [:postcode, :address_city, :self_introduction])
```

`account_update` についても同様に書けます。

masamune17 replied:
同一のメソッドを複数行で行っていた処理を一つにまとめました

betachelsea commented on app/views/books/_form.html.erb:
```diff
@@ -1,37 +1,37 @@
-  <% if book.errors.any? %>
-    <div id="error_explanation">
-      <h2><%= pluralize(book.errors.count, "error") %> prohibited this book from being saved:</h2>
+<% if book.errors.any? %>
```
総合的にインデントを一段無くした様子ですが、意図的な修正でしょうか？ 👀 
もし特別意図がない変更の場合、修正を入れない形にしたほうが `Files changed` の量を減らすことができます！

masamune17 replied:
beautifyの保存時のフォーマット機能を有効にしていたので、無効にしました。

betachelsea replied:
なるほど！そういうことでしたか、対応ありがとうございます 🙇

betachelsea commented on app/views/layouts/application.html.erb:
```diff
@@ -1,15 +1,38 @@
+<body>
+
+  <% if notice %>
+  <p class="alert alert-success"><%= notice %></p>
```
flash文言表示機能が、共通で使用できるような位置に引っ越して便利になりましたね 👍

betachelsea commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,44 @@
+
+<div class="field">
+  <%= f.label :self_introduction %><br />
+  <%= f.text_field :self_introduction, autofocus: true, autocomplete: "self_introduction" %>
```
豆知識 💡 
`f.text_area` という書き方にすると、`<input type="text"></input>`  ではなく、複数行入力扱いの `<textarea></textarea>` として出力されます！

masamune17 replied:
自己紹介文なので、機能的に"textarea"の方が合っていることから、
app/views/devise/registrations/new.html.erb
app/views/devise/registrations/edit.html.erb
のself_introduction を

f.text_field　→ f. text_area に変更いたしました。

betachelsea commented on app/views/devise/confirmations/new.html.erb:
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
```
このあたり、confirmable関係のため使っていないviewであれば、消しておくとPullRequestとしてスッキリしそうです〜。

masamune17 replied:
使用していないディレクトリを削除いたしました

app/views/devise/confirmations
app/views/devise/passwords

betachelsea commented on app/views/devise/mailer/reset_password_instructions.html.erb:
```diff
@@ -0,0 +1,8 @@
+<p>Hello <%= @resource.email %>!</p>
```
こちら、パスワードのリセットをしようとするとき送られてくるメール文面ですが、i18n対応が必要そうだと思いました。 必須要件に `Deviseが用意した画面も日本語で表示する。それ以外の新しく追加した画面もi18nで日本語化する（英語表示は考慮不要）` とあるので、メールについてもご対応お願いできますでしょうか 🙏

masamune17 replied:
パスワード再設定メールを日本語化いたしました

betachelsea commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+<ul class="users">
+  <% @users.each do |user| %>
+  <li>
+    <tr><%= link_to user.email, user %></tr>
```
`<tr>` タグですが、テーブルの行 ( table row、で tr )を表すので、いきなり `<li>` タグの直下に存在するとかなり違和感が強いです。これを実行すると、おそらくブラウザ解釈で `<tr>` は無かったことにされるのではと思いますが、ソースコード上でも `<tr>` は無しとしておいていただきたいです 🙏 

[<tr>: 表の行要素 \- HTML: HyperText Markup Language \| MDN](https://developer.mozilla.org/ja/docs/Web/HTML/Element/tr)

masamune17 replied:
app/views/users/index.html.erb の不要なタグ<tr>を削除いたしました。

betachelsea commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,14 @@
+<h2><%=t('views.users_show.title')%></h2>
+
+<tr>
+  <th><%= User.human_attribute_name(:email) %></th><br />
+  <th></th><%= @user.email %><th></th><br /><br />
+  <th><%= User.human_attribute_name(:postcode) %></th><br />
+  <th></th><%= @user.postcode %><th></th><br /><br />
+  <th><%= User.human_attribute_name(:address_city) %></th><br />
+  <th></th><%= @user.address_city %><th></th><br /><br />
+  <th><%= User.human_attribute_name(:self_introduction) %></th><br />
+  <th></th><%= @user.self_introduction %><th></th><br /><br />
+</tr>
```
こちら、 `<tr>`  や `<th>` を使っていますが、これらは `<table>` とセットで 表組みにするときに使うタグとなります。あらためてきちんと `<table>` を利用するか、もしくは方針を変えて定義リスト( `<dl>`, `<dt>`, `<dd>` ) あたりを利用するのはいかがでしょうか？
[`<dl>`: 説明リスト要素 \- HTML: HyperText Markup Language \| MDN](https://developer.mozilla.org/ja/docs/Web/HTML/Element/dl)

また、改行タグ `<br />` をレイアウトのために利用しないほうがより美しいマークアップになりますので、以降のプラクティスなどでも、レイアウト的な調整はcssに積極的に任せるように実装時に意識していただきたいです ✨

masamune17 replied:
ユーザー詳細ページをcssを用いて修飾いたしました。

以下のファイルを修正
app/views/users/show.html.erb
app/assets/stylesheets/users.scss

betachelsea commented on config/routes.rb:
```diff
@@ -1,4 +1,9 @@
 Rails.application.routes.draw do
+  get 'users/index'
+  get 'users/show'
```
これらの行の下に記載されている `resources :users, :only => [:index, :show]` が、`get 'users/index'` `get 'users/show'` と同じ意味合いとなりますので、この2行は削除しても動きそうだと思いました 👀

masamune17 replied:
不要なリクエストでしたので削除いたしました。

betachelsea commented on db/seeds.rb:
```diff
@@ -42,4 +42,17 @@ def picture_file(name)
+User.destroy_all
+
+50.times do |n|
+  name = Faker::Name.name
```
`Faker::Name` いいですね ✨

betachelsea commented on app/assets/stylesheets/users.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
このファイルは事実上、利用していない様子ですね 👀 
他にも「作成されたけど使っていない」ファイルがちらほらあるようなので、もしそれとハッキリわかるものがあれば削除をお願いしたいです 🙏

masamune17 replied:
app/views、app/assets ディレクトリの使用していないファイルを削除いたしました。

betachelsea commented on app/views/devise/sessions/new.html.erb:
```diff
@@ -0,0 +1,27 @@
+<% end %>
+
+<%= render "devise/shared/links" %>
+<a href="http://localhost:3000/letter_opener">letter_opener</a>
```
お、letter_openerのリンクがあると、開発者としては便利ですね。
ですが同時にこのような開発者向け機能を作るのはリスクでもあります。例えば、このリンクの削除を忘れてしまったまま実際のサービスとして公開されてしまうと、一般の利用者さんは「なんだこれ？」と、リンクを押しにくるでしょう。

もし開発時のみの機能を入れたい場合、忘れずに `Rails.env.development?` の判定を利用し、「開発環境でのみ表示させる」といったやり方にするのをオススメします。 `config/routes.rb` の `/letter_opener` のpathはすでに開発環境のみ有効、といった形になっているようですね。

こちら、開発環境で閲覧した時のみ、 letter_openerのリンクが表示されるように修正をお願いしたいです 🙏

masamune17 replied:
letter_openerのリンクを開発環境でのみ表示させるように変更いたしました。

<% if Rails.env.development? %>
  <%= render "devise/shared/links" %>
  <a href="http://localhost:3000/letter_opener">letter_opener</a>
<% end %>

betachelsea commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,44 @@
+<%= render "devise/shared/error_messages", resource: resource %>
+
+<div class="field">
+  <%= f.label :email %><br />
```
label と input の並びは、 `<br />` がなくてもちゃんと縦に並ぶと思うので試してみて欲しいです〜 🙏

masamune17 replied:
不要な`<br />`を削除いたしました

    app/views/devise/registrations/edit.html.erb
    app/views/devise/registrations/new.html.erb

betachelsea reviewed (commented):
コメントさせていただきました！
修正時、コメントのあった行以外にも同じ指摘を受けそうな場所も合わせて修正をお願いしたいです。不明点あればリプライいただければと思います！
( リプライは bootcamp.fjord.jp にいただけると漏れなく返信できて助かります 🙇  )

masamune17 committed:
メソッドの記述をまとめる

masamune17 committed:
自己紹介の入力フォームを変更

app/views/devise/registrations/new.html.erb
app/views/devise/registrations/edit.html.erb
のself_introduction を

f.text_field　→ f. text_area に変更いたしました。

masamune17 committed:
不要なディレクトリの削除

masamune17 committed:
パスワード再設定メールの日本語化

masamune17 committed:
不要なhtmlの要素を削除

app/views/users/index.html.erb
<tr>

masamune17 committed:
ユーザー詳細ページを修飾

masamune17 committed:
不要なリクエストを削除

masamune17 committed:
使用していないファイル、ディレクトリの削除rinosakuzyo

masamune17 committed:
letter_openerのリンクを開発環境でのみ表示させるように変更

masamune17 committed:
不要なbrタグの削除

app/views/devise/registrations/edit.html.erb
app/views/devise/registrations/new.html.erb

betachelsea reviewed (approved):
諸々細かい点、修正ありがとうございます！ 
動作も問題なかったです！ LGTM 👍
